File: cw_hrms/cw_hrms/page/user_dashboard/user_dashboard.js
frappe.pages['user-dashboard'].on_page_load = function (wrapper) {
    let page = frappe.ui.make_app_page({
        parent: wrapper,
        title: __('User Dashboard'),
        single_column: false
    });

    page.main.html('<div id="dashboard-root"></div>');
    
    create_workspace_sidebar(page);
    
    frappe.db.get_value('Employee', { user_id: frappe.session.user }, 'name', (r) => {
        let current_user_employee = r ? r.name : null;
        setup_employee_filter(page, current_user_employee);
    });
};

function setup_employee_filter(page, current_user_employee) {
    page.employee_filter = frappe.ui.form.make_control({
        parent: page.wrapper.find('.page-actions'),
        df: {
            fieldtype: 'Link',
            options: 'Employee',
            fieldname: 'employee',
            placeholder: __('Select Employee'),
            get_query: () => {
                if (!frappe.user_roles.includes('HR Manager') && !frappe.user_roles.includes('HR User')) {
                    return {
                        filters: [
                            ['Employee', 'reports_to', '=', current_user_employee, 'or'],
                            ['Employee', 'name', '=', current_user_employee]
                        ]
                    };
                }
            }
        },
        render_input: true,
    });

    // Dropdown থেকে select করলে সাথে সাথে fire হয়
    page.employee_filter.$input.on('awesomplete-selectcomplete', function() {
        let selected = page.employee_filter.get_value();
        if (selected) refresh_dashboard_data(page, selected);
    });

    // Manually type করে Enter বা blur করলে
    page.employee_filter.$input.on('change', function() {
        let selected = page.employee_filter.get_value();
        if (selected) refresh_dashboard_data(page, selected);
    });

    if (current_user_employee) {
        page.employee_filter.set_value(current_user_employee);
        refresh_dashboard_data(page, current_user_employee);
    }
}

function refresh_dashboard_data(page, employee) {
    frappe.dom.freeze(__('Loading...'));
    
    frappe.call({
        method: 'cw_hrms.cw_hrms.page.user_dashboard.user_dashboard.get_user_stats',
        args: { employee: employee },
        callback: function (r) {
            frappe.dom.unfreeze();
            if (r.message) {
                render_full_dashboard(page, r.message, employee);
            }
        }
    });
}

function render_full_dashboard(page, data, selected_employee) {
    let stats = data.stats || {};
    let leave_alloc = data.leave_allocation || [];
    let leave_hist = data.leave_history || [];
    
    let total_used_leaves = 0;
    leave_alloc.forEach(l => { total_used_leaves += flt(l.used_leave || 0); });

    let current_year = new Date().getFullYear();
    let year_options = Array.from({length: 5}, (_, i) => current_year - i)
        .map(y => `<option value="${y}" ${y === current_year ? 'selected' : ''}>${y}</option>`)
        .join('');

    let html = `
    <style>
        .dash-container { padding: 15px; background: #f8f9fc; }
        .stat-grid { display: grid; grid-template-columns: repeat(auto-fit, minmax(130px, 1fr)); gap: 15px; margin-bottom: 20px; }
        .s-card { background: white; padding: 15px; border-radius: 8px; border-bottom: 4px solid; text-align: center; box-shadow: 0 2px 4px rgba(0,0,0,0.05); }
        .s-value { font-size: 20px; font-weight: 800; }
        .main-grid { display: grid; grid-template-columns: 1.2fr 0.8fr; gap: 20px; }
        .section { background: white; padding: 15px; border-radius: 10px; box-shadow: 0 2px 8px rgba(0,0,0,0.05); margin-bottom: 20px; }
        .sec-title { font-size: 14px; font-weight: 700; margin-bottom: 12px; border-bottom: 1px solid #eee; padding-bottom: 5px; }
        .m-table { width: 100%; border-collapse: collapse; }
        .m-table td, .m-table th { padding: 8px; font-size: 12px; border-bottom: 1px solid #f4f4f4; text-align: left; }
        .badge { padding: 3px 7px; border-radius: 10px; font-size: 10px; font-weight: 600; }
        .bg-present { background: #e6fffa; color: #38a169; }
        .bg-absent { background: #fff5f5; color: #e53e3e; }
        .bg-late { background: #fffaf0; color: #d69e2e; border: 1px solid #fbd38d; }
        .bg-leave { background: #ebf8ff; color: #3182ce; }
        .bg-holiday { background: #faf5ff; color: #805ad5; }
        .bg-weekend { background: #f0fff4; color: #276749; }
        .year-select { border: 1px solid #ddd; border-radius: 6px; padding: 4px 10px; font-size: 12px; color: #555; cursor: pointer; outline: none; }
        .year-select:hover { border-color: #aaa; }
    </style>

    <div class="dash-container">

        <!-- Stat Cards -->
        <div class="stat-grid">
            ${createStatCard("Present", stats.present, "#38a169")}
            ${createStatCard("Absent", stats.absent, "#e53e3e")}
            ${createStatCard("Late", stats.late, "#d69e2e")}
            ${createStatCard("Leave", total_used_leaves, "#3182ce")}
            ${createStatCard("Holiday", (flt(stats.weekend) + flt(stats.holiday)), "#805ad5")}
        </div>

        <!-- Main Grid -->
        <div class="main-grid">

            <!-- Monthly Attendance Table -->
            <div class="section">
                <div class="sec-title">📅 Attendance Status (This Month)</div>
                <div style="max-height: 400px; overflow-y: auto;">
                    <table class="m-table">
                        <thead>
                            <tr>
                                <th>Date</th>
                                <th>In / Out</th>
                                <th>Working Hours</th>
                                <th>Status</th>
                            </tr>
                        </thead>
                        <tbody>
                            ${(data.attendance_details || []).map(a => {
                                let status = a.status;
                                let color_class = "bg-absent";

                                if (status === "Present")       color_class = "bg-present";
                                else if (status === "Late")     color_class = "bg-late";
                                else if (status === "On Leave") color_class = "bg-leave";
                                else if (status === "Holiday")  color_class = "bg-holiday";
                                else if (status === "Weekend")  color_class = "bg-weekend";

                                return `<tr>
                                    <td><b>${frappe.datetime.str_to_user(a.attendance_date)}</b></td>
                                    <td>${(a.in_time || '').split(' ')[1] || '--'} - ${(a.out_time || '').split(' ')[1] || '--'}</td>
                                    <td>${a.working_hours || '--'}</td>
                                    <td><span class="badge ${color_class}">${__(status)}</span></td>
                                </tr>`;
                            }).join('')}
                        </tbody>
                    </table>
                </div>
            </div>

            <!-- Right Side -->
            <div>
                <!-- Leave Balance -->
                <div class="section">
                    <div class="sec-title">📊 Leave Balance (Unused / Total)</div>
                    <table class="m-table">
                        <tbody>
                            ${leave_alloc.length ? leave_alloc.map(l => `
                                <tr>
                                    <td>${l.leave_type}</td>
                                    <td align="right"><strong>${l.unused_leaves}</strong> / ${l.total_allocated}</td>
                                </tr>
                            `).join('') : '<tr><td colspan="2" align="center">No Records</td></tr>'}
                        </tbody>
                    </table>
                </div>

                <!-- Leave History -->
                <div class="section">
                    <div class="sec-title">🗓️ Recent Leave History</div>
                    <table class="m-table">
                        <tbody>
                            ${leave_hist.length ? leave_hist.map(l => `
                                <tr>
                                    <td>${frappe.datetime.str_to_user(l.from_date)}</td>
                                    <td><strong>${l.total_leave_days} d</strong></td>
                                    <td><span class="badge" style="background:#e2e5ff; color:#5e72e4;">${l.leave_type}</span></td>
                                </tr>
                            `).join('') : '<tr><td colspan="3" align="center">No Records</td></tr>'}
                        </tbody>
                    </table>
                </div>
            </div>
        </div>

        <!-- Yearly Chart -->
        <div class="section">
            <div class="sec-title" style="display:flex; justify-content:space-between; align-items:center;">
                <span id="chartTitle">📈 Attendance Overview - ${current_year}</span>
                <select id="yearSelect" class="year-select">
                    ${year_options}
                </select>
            </div>
            <div style="height:280px;"><canvas id="attBarChart"></canvas></div>
        </div>

    </div>`;

    page.wrapper.find('#dashboard-root').html(html);

    // Year filter event
    page.wrapper.find('#yearSelect').on('change', function() {
        let selected_year = parseInt($(this).val());
        page.wrapper.find('#chartTitle').text(`📈 Attendance Yearly Chart - ${selected_year}`);
        render_chart(selected_employee, selected_year);
    });

    frappe.require('https://cdnjs.cloudflare.com/ajax/libs/Chart.js/4.4.1/chart.umd.js', function() {
        render_chart(selected_employee, current_year);
    });
}

function render_chart(employee, year) {
    frappe.call({
        method: 'cw_hrms.cw_hrms.page.user_dashboard.user_dashboard.get_yearly_attendance',
        args: { year: year || new Date().getFullYear(), employee: employee },
        callback: function(r) {
            let rows = r.message || [];
            let canvas = document.getElementById('attBarChart');
            if (!canvas) return;
            let ctx = canvas.getContext('2d');
            if (window.dashboardChart) window.dashboardChart.destroy();
            window.dashboardChart = new Chart(ctx, {
                type: 'bar',
                data: {
                    labels: rows.map(d => d.month),
                    datasets: [
                        { label: 'Present', data: rows.map(d => d.present), backgroundColor: '#38a169', stack: 's' },
                        { label: 'Absent',  data: rows.map(d => d.absent),  backgroundColor: '#e53e3e', stack: 's' },
                        { label: 'Late',    data: rows.map(d => d.late),    backgroundColor: '#d69e2e', stack: 's' },
                        { label: 'Leave',   data: rows.map(d => d.leave),   backgroundColor: '#3182ce', stack: 's' }
                    ]
                },
                options: {
                    responsive: true,
                    maintainAspectRatio: false,
                    scales: {
                        x: { stacked: true },
                        y: { stacked: true, beginAtZero: true }
                    },
                    plugins: {
                        tooltip: {
                            callbacks: {
                                label: function(context) {
                                    let dataIndex = context.dataIndex;
                                    let present = rows[dataIndex].present || 0;
                                    let absent  = rows[dataIndex].absent  || 0;
                                    let late    = rows[dataIndex].late    || 0;
                                    let leave   = rows[dataIndex].leave   || 0;
                                    let total   = present + absent + late + leave;

                                    let value   = context.parsed.y || 0;
                                    let percent = total > 0 ? ((value / total) * 100).toFixed(1) : 0;

                                    return ` ${context.dataset.label}: ${value} (${percent}%)`;
                                }
                            }
                        }
                    }
                }
            });
        }
    });
}

function createStatCard(label, val, color) {
    return `<div class="s-card" style="border-bottom-color: ${color}">
        <div style="font-size: 11px; color: #888;">${__(label)}</div>
        <div class="s-value" style="color: ${color}">${val || 0}</div>
    </div>`;
}

// সাইডবার ফাংশন
function create_workspace_sidebar(page) {
    let list_sidebar = $(`
        <div class="list-sidebar overlay-sidebar hidden-xs hidden-sm">
            <div class="desk-sidebar list-unstyled sidebar-menu"></div>
        </div>
    `).appendTo(page.sidebar);

    let sidebar = list_sidebar.find('.desk-sidebar');

    frappe.xcall('frappe.desk.desktop.get_workspace_sidebar_items').then((data) => {
        if (!data || !data.pages) return;
        let public_pages  = data.pages.filter(p => p.public);
        let private_pages = data.pages.filter(p => !p.public);

        if (public_pages.length > 0)  build_sidebar_section(sidebar, 'Public',   public_pages,  true);
        if (private_pages.length > 0) build_sidebar_section(sidebar, 'Personal', private_pages, false);
    });
}

function build_sidebar_section(sidebar, title, pages, is_public) {
    let root_pages = pages.filter(p => !p.parent_page);
    let section = $(`
        <div class="standard-sidebar-section nested-container" data-title="${title}">
            <button class="btn-reset standard-sidebar-label">
                <span>${frappe.utils.icon('es-line-down', 'xs')}</span>
                <span class="section-title">${__(title)}</span>
            </button>
        </div>
    `).appendTo(sidebar);

    root_pages.forEach(page => {
        append_sidebar_item(section, page, pages, is_public);
    });
}

function append_sidebar_item(container, item, all_pages, is_public) {
    let route = is_public
        ? frappe.router.slug(item.title)
        : 'private/' + frappe.router.slug(item.title);

    let $item = $(`
        <div class="sidebar-item-container" item-name="${item.title}">
            <div class="desk-sidebar-item standard-sidebar-item">
                <a href="/app/${route}" class="item-anchor">
                    <span class="sidebar-item-icon">
                        ${is_public
                            ? frappe.utils.icon(item.icon || 'folder-normal', 'md')
                            : `<span class="indicator ${item.indicator_color || 'blue'}"></span>`}
                    </span>
                    <span class="sidebar-item-label">${__(item.title)}</span>
                </a>
            </div>
            <div class="sidebar-child-item nested-container hidden"></div>
        </div>
    `).appendTo(container);

    let child_items = all_pages.filter(p => p.parent_page == item.title);
    if (child_items.length > 0) {
        let child_container = $item.find('.sidebar-child-item');
        child_items.forEach(child => append_sidebar_item(child_container, child, all_pages, is_public));
        child_container.removeClass('hidden');
    }
}

File: cw_hrms/cw_hrms/page/user_dashboard/user_dashboard.py
import frappe
from frappe.utils import nowdate, get_first_day, getdate, add_days, flt

@frappe.whitelist()
def get_user_stats(employee=None):
    user = frappe.session.user
    if not employee:
        employee = frappe.db.get_value("Employee", {"user_id": user}, "name")
    
    employee_doc = frappe.db.get_value("Employee", {"name": employee}, ["name", "company", "holiday_list"], as_dict=True)
    
    if not employee_doc:
        return {"stats": {}, "attendance_details": [], "leave_allocation": [], "leave_history": []}
    
    employee = employee_doc.name
    today = getdate(nowdate())
    yesterday = add_days(today, -1)
    month_start = getdate(get_first_day(today))
    year_start = f"{today.year}-01-01"
    year_end = f"{today.year}-12-31"

    data = {
        "stats": {"present": 0, "absent": 0, "late": 0, "leave": 0, "holiday": 0, "weekend": 0, "home_office": 0},
        "leave_allocation": [],
        "attendance_details": [],
        "leave_history": []
    }

    # ১. এটেনডেন্স ডাটা
    attendance_records = frappe.get_all("Attendance",
        filters={"employee": employee, "attendance_date": ["between", [month_start, yesterday]], "docstatus": 1},
        fields=["attendance_date", "status", "late_entry", "in_time", "out_time", "working_hours"]
    )
    att_dict = {getdate(d.attendance_date): d for d in attendance_records}

    # ২. হলিডে ডাটা
    h_list = employee_doc.holiday_list or frappe.db.get_value("Company", employee_doc.company, "default_holiday_list")
    holiday_dict = {}
    if h_list:
        holidays = frappe.db.sql("""SELECT holiday_date, description, weekly_off FROM `tabHoliday`
                                    WHERE parent = %s AND holiday_date BETWEEN %s AND %s""",
                                 (h_list, month_start, yesterday), as_dict=True)
        holiday_dict = {getdate(h.holiday_date): h for h in holidays}

    # ৩. Working Hours Format
    def format_to_hh_mm_ss(hours_float):
        if not hours_float: return "00:00:00"
        total_seconds = int(flt(hours_float) * 3600)
        hh = total_seconds // 3600
        mm = (total_seconds % 3600) // 60
        ss = total_seconds % 60
        return f"{hh:02d}:{mm:02d}:{ss:02d}"

    # ৪. Attendance Processing Loop
    curr_date = month_start
    temp_details = []

    while curr_date <= yesterday:
        status_to_show, in_t, out_t, work_h = "Absent", None, None, 0

        if curr_date in holiday_dict:
            h_info = holiday_dict[curr_date]
            if curr_date in att_dict and att_dict[curr_date].status == "Present":
                att = att_dict[curr_date]
                in_t, out_t, work_h = att.in_time, att.out_time, att.working_hours
                if att.late_entry:
                    status_to_show = "Late"
                    data["stats"]["present"] += 1
                    data["stats"]["late"] += 1
                else:
                    status_to_show = "Present"
                    data["stats"]["present"] += 1
            else:
                status_to_show = "Weekend" if h_info.weekly_off else (h_info.description or "Holiday")
                data["stats"]["weekend" if h_info.weekly_off else "holiday"] += 1

        elif curr_date in att_dict:
            att = att_dict[curr_date]
            in_t, out_t, work_h = att.in_time, att.out_time, att.working_hours

            if att.status == "Present" and att.late_entry:
                # Late হলে present count এ থাকবে + late আলাদা count
                status_to_show = "Late"
                data["stats"]["present"] += 1
                data["stats"]["late"] += 1
            else:
                status_to_show = att.status
                key = status_to_show.lower().replace(" ", "_")
                if key == "on_leave":
                    key = "leave"
                if key in data["stats"]:
                    data["stats"][key] += 1

        else:
            data["stats"]["absent"] += 1

        temp_details.append({
            "attendance_date": curr_date,
            "status": status_to_show,
            "in_time": str(in_t) if in_t else None,
            "out_time": str(out_t) if out_t else None,
            "working_hours": format_to_hh_mm_ss(work_h)
        })
        curr_date = add_days(curr_date, 1)

    data["attendance_details"] = sorted(temp_details, key=lambda x: x['attendance_date'], reverse=True)

    # ৫. লিভ Allocation & Calculation
    allocations = frappe.get_all("Leave Allocation",
        filters={"employee": employee, "docstatus": 1, "from_date": ["<=", today], "to_date": [">=", today]},
        fields=["leave_type", "total_leaves_allocated"])

    leave_data = []
    for alloc in allocations:
        used_leaves = frappe.db.get_value("Leave Application", {
            "employee": employee,
            "leave_type": alloc.leave_type,
            "status": "Approved",
            "docstatus": 1,
            "from_date": [">=", year_start],
            "to_date": ["<=", year_end]
        }, "sum(total_leave_days)") or 0

        remaining = flt(alloc.total_leaves_allocated) - flt(used_leaves)
        leave_data.append({
            "leave_type": alloc.leave_type,
            "total_allocated": alloc.total_leaves_allocated,
            "unused_leaves": remaining,
            "used_leave": flt(used_leaves)
        })
    data["leave_allocation"] = leave_data

    # ৬. লিভ হিস্ট্রি
    data["leave_history"] = frappe.get_all("Leave Application",
        filters={"employee": employee, "status": "Approved", "docstatus": 1, "from_date": [">=", year_start]},
        fields=["leave_type", "from_date", "to_date", "total_leave_days"],
        order_by="from_date desc")

    return data

@frappe.whitelist()
def get_yearly_attendance(year=None, employee=None):
    user = frappe.session.user

    if not employee:
        employee = frappe.db.get_value("Employee", {"user_id": user}, "name")

    if not employee:
        return []

    if not year:
        year = getdate(nowdate()).year

    year = int(year)
    today = getdate(nowdate())
    yesterday = add_days(today, -1)
    year_start = getdate(f"{year}-01-01")
    year_end = getdate(f"{year}-12-31")

    # Future year হলে সব 0
    if year_start > today:
        return [
            {
                "month": ["Jan","Feb","Mar","Apr","May","Jun","Jul","Aug","Sep","Oct","Nov","Dec"][i],
                "present": 0, "absent": 0, "leave": 0, "late": 0
            }
            for i in range(12)
        ]

    # Employee info আনো
    employee_doc = frappe.db.get_value("Employee", {"name": employee},
        ["company", "holiday_list", "date_of_joining", "relieving_date"], as_dict=True)

    if not employee_doc:
        return []

    # এই employee এর সর্বপ্রথম attendance record কোনদিন
    first_attendance = frappe.db.get_value("Attendance",
        {"employee": employee, "docstatus": 1},
        "min(attendance_date)")

    joining_date = getdate(employee_doc.date_of_joining) if employee_doc.date_of_joining else None
    first_att_date = getdate(first_attendance) if first_attendance else None

    # loop_start নির্ধারণ — তিনটার মধ্যে সবচেয়ে কার্যকর তারিখ
    # joining date আর first attendance date এর মধ্যে যেটা পরে সেটা নাও
    # তারপর সেটা year_start এর সাথে তুলনা করে যেটা পরে সেটা নাও
    if first_att_date and joining_date:
        effective_start = max(joining_date, first_att_date)
    elif first_att_date:
        effective_start = first_att_date
    elif joining_date:
        effective_start = joining_date
    else:
        effective_start = year_start

    loop_start = max(year_start, effective_start)

    # loop_end নির্ধারণ
    relieving_date = getdate(employee_doc.relieving_date) if employee_doc.relieving_date else None
    loop_end = min(yesterday, year_end)
    if relieving_date:
        loop_end = min(loop_end, relieving_date)

    # Selected year এ কোনো কার্যকর দিন নেই
    if loop_start > loop_end:
        return [
            {
                "month": ["Jan","Feb","Mar","Apr","May","Jun","Jul","Aug","Sep","Oct","Nov","Dec"][i],
                "present": 0, "absent": 0, "leave": 0, "late": 0
            }
            for i in range(12)
        ]

    # Attendance আনো
    records = frappe.get_all("Attendance",
        filters={
            "employee": employee,
            "attendance_date": ["between", [loop_start, loop_end]],
            "docstatus": 1
        },
        fields=["attendance_date", "status", "late_entry"]
    )
    att_dict = {getdate(a.attendance_date): a for a in records}

    # Holiday আনো
    h_list = employee_doc.holiday_list or frappe.db.get_value("Company", employee_doc.company, "default_holiday_list")

    holiday_set = set()
    if h_list:
        holidays = frappe.db.sql("""SELECT holiday_date FROM `tabHoliday`
                                    WHERE parent = %s AND holiday_date BETWEEN %s AND %s""",
                                 (h_list, loop_start, loop_end), as_dict=True)
        holiday_set = {getdate(h.holiday_date) for h in holidays}

    monthly = {i: {"present": 0, "absent": 0, "leave": 0, "late": 0} for i in range(12)}

    curr_date = loop_start
    while curr_date <= loop_end:
        m = curr_date.month - 1

        if curr_date in holiday_set:
            if curr_date in att_dict and att_dict[curr_date].status == "Present":
                monthly[m]["present"] += 1
                if att_dict[curr_date].late_entry:
                    monthly[m]["late"] += 1

        elif curr_date in att_dict:
            a = att_dict[curr_date]
            s = (a.status or "").lower().replace(" ", "_")

            if "present" in s:
                monthly[m]["present"] += 1
            elif "absent" in s:
                monthly[m]["absent"] += 1
            elif "leave" in s or "on_leave" in s:
                monthly[m]["leave"] += 1

            if a.late_entry:
                monthly[m]["late"] += 1

        else:
            monthly[m]["absent"] += 1

        curr_date = add_days(curr_date, 1)

    return [
        {
            "month": ["Jan","Feb","Mar","Apr","May","Jun","Jul","Aug","Sep","Oct","Nov","Dec"][i],
            "present": monthly[i]["present"],
            "absent":  monthly[i]["absent"],
            "leave":   monthly[i]["leave"],
            "late":    monthly[i]["late"]
        }
        for i in range(12)
    ]
@frappe.whitelist()
def get_employee_list():
    if "System Manager" not in frappe.get_roles() and "HR Manager" not in frappe.get_roles():
        return []

    return frappe.get_all("Employee",
        filters={"status": "Active"},
        fields=["name", "employee_name"],
        order_by="employee_name asc"
    )